### ISSUE TITLE
Negative searching

### ISSUE THREAD

**Issue by sanmadjack** [shish/shimmie2#666]
I found an issue with negative searching on PGSQL. Searching for one positive tag and one negative tag produces this query:

```
SELECT images.* 
FROM images 
WHERE images.id IN 
( SELECT image_id 
FROM image_tags 
WHERE tag_id IN (28) AND tag_id NOT IN (386) 
GROUP BY image_id HAVING COUNT(image_id) >= :search_score )
```

This returns all images matching the positive tag, but none with the negative tag are omitted.  It looks like this was implemented when the search_score idea (which is very clever) was implemented to reduce the number of subqueries. Unless I'm missing something (which I'm pretty sure I'm not, since I found this because negative searches weren't working), this needs to be broken into two subqueries, one IN and one NOT IN.

I've got a commit ready that does this (which has resolved my searching issues), but it got me thinking about negative behavior. Positive tags are searched in an AND fashion, all positive tags must be found. Should negative tags follow the same rules? If I search for A, -B, -C, should I see results that are AB and AC, only filtering out ABC? Or should they be in an OR fashion, filtering out AB, AC, and ABC? It seems to me the more natural thinking is for it to be an OR implementation, but I'm not sure. Is there an OR operator in shimmie that I'm not aware of?

**Comment by sanmadjack** [comment]
Also congrats, 666 issues.

**Comment by shish** [comment]
The behaviour most people seem to expect is to return results with all of the positive tags, and none of the negative tags

Other considerations:
- Don't melt the database
- Handle wildcards somehow (I think OR makes sense here - so "A* B -C" would mean "(Alice OR Aardvark OR Ampersand) AND B AND (NOT C)")
- Bonus points for working efficiently on all databases (historically mysql has been awful at subqueries, eg `select * from images where image_id in (select 42)` - postgres runs the subquery once, sees that the result is a constant 42, and then does an index lookup to jump straight to image 42. MySQL does a sequential scan over all images, running the subquery once for every row... I don't know if it's still that bad, I've managed to avoid it for the past few years :P)

Except for wildcards acting in a very limited kind-of-OR way, all the searching is currently AND / AND NOT

(Also in addition to the "SQL search for good databases" and "SQL search for mysql", there is the search accelerator which keeps all the data in RAM and does a very limited search (no wildcards, no metadata queries) very quickly)

**Comment by sanmadjack** [comment]
I did some tests using different methods of performing the queries, focusing on three methods:

1. IN and NOT IN clauses 
2. JOINS
3. JOIN and INTERSECT/EXCEPT

Looking at the data I found that three kinds of subqueries would be needed:

1. A subquery for all single positive tags, using the search_score trick. These need to be all treated in a strict AND fashion.
2. A subquery for each wildcard that resolved into multiple tags. If a wildcard resolved to only one tag, then it could be lumped in with the other single tags in #1, but if it resolved to multiple tags those need to be treated in an OR manner amongst themselves and in an AND fashion in regards to the other criteria. This means more wildcards = more subqueries.
3. A subquery for all negative tags, wildcard or otherwise. All negative tags would be treated in an OR fashion, so they can be lumped into one subquery similar to #2, but using a NOT IN or other such excluder.

I tested several patterns against my database. I'm running these queries against a database with over 600,000 images and 39,000 tags. It's not the largest sample size, but I think it's decent enough to get an idea of what higher-end performance would be like. 

NOTE: I am not a db admin. I am a software developer who has happened to do a lot of work with databases, and optimized many queries. My understanding of database mechanics is not perfect and is largely gleaned through observation.

First using IN/NOT IN similar to the manner shown in the documentation:
```
    SELECT images.* 
FROM images 
WHERE images.id IN ( SELECT image_id FROM image_tags WHERE tag_id IN (2082) GROUP BY image_id HAVING COUNT(image_id) >= 1 ) 
AND images.id IN ( SELECT image_id FROM image_tags WHERE tag_id IN (6613, 6619) GROUP BY image_id ) 
AND images.id IN ( SELECT image_id FROM image_tags WHERE tag_id IN (14318, 7346, 8631, 26382, 30742, 13562, 5267, 9558, 26352, 5103, 13149, 2083) GROUP BY image_id ) 
AND images.id NOT IN ( SELECT image_id FROM image_tags WHERE tag_id IN (386) GROUP BY image_id )
```
This query is basically the worst of those tested. It never returns in less than 400ms, and the execution plan takes 3 times as long (1.5ms)  to form as the best performing one. One key thing is that the execution plan delayed looking up from the images table until it ran the first sub-query, meaning that it would filter down the potential image set based on that first query, and then would check each image against the subsequent queries. This would indicate that this solution would not scale up very well with more criteria.

So, I tried using joins instead. This means that the subqueries are not necessarily strict subqueries, as the database optimizer will attempt to be smart and sub-query only on related data.

```
    SELECT images.*
    FROM images
        INNER JOIN ( SELECT image_id FROM image_tags
            WHERE tag_id IN (2082) GROUP BY image_id HAVING COUNT(image_id) >= 1 ) pos1 ON pos1.image_id = images.id
        INNER JOIN ( SELECT image_id FROM image_tags WHERE tag_id IN (6613, 6619) GROUP BY image_id ) pos2 ON pos2.image_id = images.id
        INNER JOIN ( SELECT image_id FROM image_tags
            WHERE tag_id IN (14318, 7346, 8631, 26382, 30742, 13562, 5267, 9558, 26352, 5103, 13149, 2083) GROUP BY image_id ) pos3 ON pos3.image_id = images.id
        LEFT JOIN ( SELECT image_id FROM image_tags
            WHERE tag_id IN (386) GROUP BY image_id ) neg ON neg.image_id = images.id WHERE neg.image_id IS NULL
```
This returns in about 10ms, significantly faster than the IN-based query. A big difference is that the LEFT JOIN/IS NULL pattern allows the database to make use of a NOT JOIN optimization internally. Using joins in general let the database optimize much more effectively, but the execution plan is still filtering down the images table only after the first subquery, so any subsequent joins would still have a performance impact. That said, it runs fast, and the execution plan generates in only 1 ms.

But that wasn't good enough. I did some more research into these kinds of queries and came up with another query pattern using INTERSECT and EXCEPT:

```
    SELECT images.*
    FROM images INNER JOIN (
        SELECT image_id FROM image_tags
            WHERE tag_id IN (2082) GROUP BY image_id HAVING COUNT(image_id) >= 1
        INTERSECT
        SELECT image_id FROM image_tags WHERE tag_id IN (6613, 6619)
        INTERSECT
        SELECT image_id FROM image_tags
            WHERE tag_id IN (14318, 7346, 8631, 26382, 30742, 13562, 5267, 9558, 26352, 5103, 13149, 2083)
        EXCEPT
        SELECT image_id FROM image_tags
            WHERE tag_id IN (386)
    ) tags ON tags.image_id = images.id;
```
Basically our subquery completely calculates the images qualifying for all tags all at once, leaving one join to the images table. This query returns in about the same 10ms as the join query, but I think that's just the minimum amount of time any query could return from my server at present. The real interesting information is in the execution plan:

```
Nested Loop  (cost=506.22..6920.81 rows=44 width=636) (actual time=4.547..4.574 rows=5 loops=1)
  ->  Subquery Scan on tags  (cost=505.79..6549.34 rows=44 width=4) (actual time=4.530..4.533 rows=5 loops=1)
        ->  HashSetOp Except  (cost=505.79..6548.90 rows=44 width=8) (actual time=4.529..4.530 rows=5 loops=1)
              ->  Append  (cost=505.79..6548.46 rows=177 width=8) (actual time=4.182..4.486 rows=298 loops=1)
                    ->  Result  (cost=505.79..6045.15 rows=44 width=8) (actual time=4.181..4.183 rows=5 loops=1)
                          ->  HashSetOp Intersect  (cost=505.79..6044.71 rows=44 width=8) (actual time=4.180..4.181 rows=5 loops=1)
                                ->  Append  (cost=505.79..6040.62 rows=1636 width=8) (actual time=3.253..4.118 rows=332 loops=1)
                                      ->  Result  (cost=505.79..1471.68 rows=44 width=8) (actual time=3.252..3.258 rows=5 loops=1)
                                            ->  HashSetOp Intersect  (cost=505.79..1471.24 rows=44 width=8) (actual time=3.251..3.255 rows=5 loops=1)
                                                  ->  Append  (cost=505.79..1470.47 rows=309 width=8) (actual time=0.555..2.869 rows=1665 loops=1)
                                                        ->  Subquery Scan on "*SELECT* 1"  (cost=505.79..508.89 rows=44 width=8) (actual time=0.555..0.765 rows=207 loops=1)
                                                              ->  GroupAggregate  (cost=505.79..508.45 rows=44 width=4) (actual time=0.553..0.715 rows=207 loops=1)
                                                                    Group Key: image_tags.image_id
                                                                    Filter: (count(image_tags.image_id) >= 1)
                                                                    ->  Sort  (cost=505.79..506.12 rows=133 width=4) (actual time=0.544..0.568 rows=207 loops=1)
                                                                          Sort Key: image_tags.image_id
                                                                          Sort Method: quicksort  Memory: 34kB
                                                                          ->  Bitmap Heap Scan on image_tags  (cost=9.46..501.10 rows=133 width=4) (actual time=0.089..0.455 rows=207 loops=1)
                                                                                Recheck Cond: (tag_id = 2082)
                                                                                Heap Blocks: exact=204
                                                                                ->  Bitmap Index Scan on images_tags_tag_id_idx  (cost=0.00..9.43 rows=133 width=0) (actual time=0.049..0.049 rows=228 loops=1)
                                                                                      Index Cond: (tag_id = 2082)
                                                        ->  Subquery Scan on "*SELECT* 2"  (cost=18.92..960.03 rows=265 width=8) (actual time=0.324..1.845 rows=1458 loops=1)
                                                              ->  Bitmap Heap Scan on image_tags image_tags_1  (cost=18.92..957.38 rows=265 width=4) (actual time=0.323..1.510 rows=1458 loops=1)
                                                                    Recheck Cond: (tag_id = ANY ('{6613,6619}'::integer[]))
                                                                    Heap Blocks: exact=537
                                                                    ->  Bitmap Index Scan on images_tags_tag_id_idx  (cost=0.00..18.85 rows=265 width=0) (actual time=0.217..0.217 rows=1469 loops=1)
                                                                          Index Cond: (tag_id = ANY ('{6613,6619}'::integer[]))
                                      ->  Subquery Scan on "*SELECT* 3"  (cost=113.50..4560.76 rows=1592 width=8) (actual time=0.174..0.802 rows=327 loops=1)
                                            ->  Bitmap Heap Scan on image_tags image_tags_2  (cost=113.50..4544.84 rows=1592 width=4) (actual time=0.173..0.720 rows=327 loops=1)
                                                  Recheck Cond: (tag_id = ANY ('{14318,7346,8631,26382,30742,13562,5267,9558,26352,5103,13149,2083}'::integer[]))
                                                  Heap Blocks: exact=317
                                                  ->  Bitmap Index Scan on images_tags_tag_id_idx  (cost=0.00..113.10 rows=1592 width=0) (actual time=0.115..0.115 rows=351 loops=1)
                                                        Index Cond: (tag_id = ANY ('{14318,7346,8631,26382,30742,13562,5267,9558,26352,5103,13149,2083}'::integer[]))
                    ->  Subquery Scan on "*SELECT* 4"  (cost=9.46..502.43 rows=133 width=8) (actual time=0.046..0.256 rows=293 loops=1)
                          ->  Bitmap Heap Scan on image_tags image_tags_3  (cost=9.46..501.10 rows=133 width=4) (actual time=0.045..0.188 rows=293 loops=1)
                                Recheck Cond: (tag_id = 386)
                                Heap Blocks: exact=34
                                ->  Bitmap Index Scan on images_tags_tag_id_idx  (cost=0.00..9.43 rows=133 width=0) (actual time=0.036..0.036 rows=324 loops=1)
                                      Index Cond: (tag_id = 386)
  ->  Index Scan using images_pkey on images  (cost=0.42..8.44 rows=1 width=636) (actual time=0.006..0.006 rows=1 loops=5)
        Index Cond: (id = tags.image_id)
```
Some specific notes:
1. Operation costs. The other queries are above 9,000 at least, and this reports 6920. I'm simplifying since costs are a little more complex than just comparing them, but in general smaller numbers are still better.
3. All of the image_tag operations are handled only comparing against each other. The comparison is quick, well-optimized in pgsql, and ends with the absolute final list of qualifying items without any further filtering needed.
2. The scan of the images table doesn't happen until the very end of the query. This means the number of records in that table has a negligible impact on the performance, since it's not doing any per-image checks. The cost for the scan of images in the first query is over 400, and the second query has the same as the third, a cost of 8.44 (because it's filtered down the possible items better). The big difference is that this one happens at the end, so that there are no subsequent criteria to evaluate on each row. 

I found that removing the GROUP BY expressions from the last 3 subqueries made the query more efficient, since the database was able to better optimize it when just working against the table indexes without the grouping operation getting in-between. So, I tried again with no grouping at all:

```
EXPLAIN ANALYZE
SELECT images.*
FROM images INNER JOIN (
    SELECT image_id FROM image_tags WHERE tag_id = 2082
        INTERSECT
        SELECT image_id FROM image_tags WHERE tag_id IN (6613, 6619)
        INTERSECT
        SELECT image_id FROM image_tags
        WHERE tag_id IN (14318, 7346, 8631, 26382, 30742, 13562, 5267, 9558, 26352, 5103, 13149, 2083)
        EXCEPT
        SELECT image_id FROM image_tags
        WHERE tag_id IN (386)
) tags ON tags.image_id = images.id;
```
Returned in 4 ms, best time by far. First-row costs went from 506.22..6920.81 to 9.88..7654.41. A slightly higher cost, but the smaller starting number indicates that the inner queries are likely to perform much more efficiently.

But, as complex as I made this query, it is only using one positive single-tag criteria. I added a second positive tag to the INNER JOIN query and its cost changed to 7042.80..7792.40, a lower start and cost, presumably because of the initial subquery further reducing the potential data set. 

Adding the same criteria to the INTERSECT/EXCLUDE criteria using a group by on the first subquery resulted in a cost of 959.13..7740.60 . That's a interesting jump from the 506.22..6920.8, indicating the subquery cost went up as the the group by gained another criteria, but since the maximum data set is then reduced, it's possible it would improve performance.  This would indicate that the minimum cost of this query will directly increase with every criteria, but overall still not as much as the other queries.

So, I then added the same criteria to the INTERSECT/EXCLUDE without using a group by, resulting in this:
```

    SELECT images.*
    FROM images INNER JOIN (
        SELECT image_id FROM image_tags WHERE tag_id = 2082
        INTERSECT
            SELECT image_id FROM image_tags WHERE tag_id = 6619
        INTERSECT
        SELECT image_id FROM image_tags WHERE tag_id IN (6613, 6619)
        INTERSECT
        SELECT image_id FROM image_tags
            WHERE tag_id IN (14318, 7346, 8631, 26382, 30742, 13562, 5267, 9558, 26352, 5103, 13149, 2083)
        EXCEPT
        SELECT image_id FROM image_tags
            WHERE tag_id IN (386)
    ) tags ON tags.image_id = images.id;
```
This resulted in a cost of 9.88..8160.16. Based on the explain, each subquery with a single tag increased the cost by 500. So, one more positive tag criteria, for 3 total. GROUPED the first-line cost is 1391.63..8549.49 and ungrouped it's 9.88..8665.92. GROUPED BY's start continues to drop with more criteria, but the cost is increasing at about 800 per criteria added. 

So, it seems to me that the ungrouped INTERSECT/EXCEPT query is the best choice. The cost increase per criteria is smaller, the start never budges, the delayed images images join is great with larger databases, and so far in my tests it has always returned the fastest.

Once again, not a database expert, and I learned a whole lot of this in the last 24 hours. I'd be very happy for someone to critique this analysis and let me know if I'm off-base about something.

**Comment by sanmadjack** [comment]
I found one more thing. There are currently these indexes on the image_tags table:

1. image_id
2. tag_id
3. image_id, tag_id

So, index order matters. All of the tag-based search criteria start their lookups with tag_id, which means that whatever index it uses will likely have tag_id first. The explains for the various queries I tried above confirm this, the only index used by all of these queries is the one that is just tag_id.

This is okay for the most part, but since we then also join based on image_id, both to the images table and back into the image_tags table these queries would be able to take advantage of a multi-column index that is more suited for it. 

So, I tried making an additional index on tag_id, image_id (unique, of course), here's the analyze from my recommended query after that:

```
Nested Loop  (cost=0.85..8114.32 rows=143 width=636) (actual time=4.039..4.068 rows=5 loops=1)
  ->  Subquery Scan on tags  (cost=0.43..6927.04 rows=143 width=4) (actual time=4.024..4.028 rows=5 loops=1)
        ->  HashSetOp Except  (cost=0.43..6925.61 rows=143 width=8) (actual time=4.023..4.025 rows=5 loops=1)
              ->  Append  (cost=0.43..6924.89 rows=286 width=8) (actual time=3.726..3.977 rows=298 loops=1)
                    ->  Result  (cost=0.43..6495.75 rows=143 width=8) (actual time=3.725..3.728 rows=5 loops=1)
                          ->  HashSetOp Intersect  (cost=0.43..6494.32 rows=143 width=8) (actual time=3.723..3.726 rows=5 loops=1)
                                ->  Append  (cost=0.43..6489.66 rows=1864 width=8) (actual time=3.055..3.660 rows=351 loops=1)
                                      ->  Result  (cost=0.43..1707.13 rows=143 width=8) (actual time=3.054..3.058 rows=5 loops=1)
                                            ->  HashSetOp Intersect  (cost=0.43..1705.70 rows=143 width=8) (actual time=3.053..3.056 rows=5 loops=1)
                                                  ->  Append  (cost=0.43..1704.63 rows=430 width=8) (actual time=1.185..2.827 rows=1476 loops=1)
                                                        ->  Result  (cost=0.43..859.01 rows=143 width=8) (actual time=1.185..1.190 rows=5 loops=1)
                                                              ->  HashSetOp Intersect  (cost=0.43..857.58 rows=143 width=8) (actual time=1.184..1.187 rows=5 loops=1)
                                                                    ->  Append  (cost=0.43..856.86 rows=286 width=8) (actual time=0.042..0.991 rows=697 loops=1)
                                                                          ->  Subquery Scan on "*SELECT* 1"  (cost=0.43..427.72 rows=143 width=8) (actual time=0.041..0.350 rows=221 loops=1)
                                                                                ->  Index Only Scan using image_tags_tag_id_image_id_uindex on image_tags  (cost=0.43..426.29 rows=143 width=4) (actual time=0.040..0.294 rows=221 loops=1)
                                                                                      Index Cond: (tag_id = 2082)
                                                                                      Heap Fetches: 202
                                                                          ->  Subquery Scan on "*SELECT* 2"  (cost=0.43..427.72 rows=143 width=8) (actual time=0.022..0.523 rows=476 loops=1)
                                                                                ->  Index Only Scan using image_tags_tag_id_image_id_uindex on image_tags image_tags_1  (cost=0.43..426.29 rows=143 width=4) (actual time=0.021..0.407 rows=476 loops=1)
                                                                                      Index Cond: (tag_id = 6619)
                                                                                      Heap Fetches: 431
                                                        ->  Subquery Scan on "*SELECT* 3"  (cost=0.43..843.47 rows=287 width=8) (actual time=0.021..1.387 rows=1471 loops=1)
                                                              ->  Index Only Scan using image_tags_tag_id_image_id_uindex on image_tags image_tags_2  (cost=0.43..840.60 rows=287 width=4) (actual time=0.021..1.031 rows=1471 loops=1)
                                                                    Index Cond: (tag_id = ANY ('{6613,6619}'::integer[]))
                                                                    Heap Fetches: 1311
                                      ->  Subquery Scan on "*SELECT* 4"  (cost=0.43..4773.20 rows=1721 width=8) (actual time=0.017..0.539 rows=346 loops=1)
                                            ->  Index Only Scan using image_tags_tag_id_image_id_uindex on image_tags image_tags_3  (cost=0.43..4755.99 rows=1721 width=4) (actual time=0.017..0.451 rows=346 loops=1)
                                                  Index Cond: (tag_id = ANY ('{14318,7346,8631,26382,30742,13562,5267,9558,26352,5103,13149,2083}'::integer[]))
                                                  Heap Fetches: 313
                    ->  Subquery Scan on "*SELECT* 5"  (cost=0.43..427.72 rows=143 width=8) (actual time=0.022..0.199 rows=293 loops=1)
                          ->  Index Only Scan using image_tags_tag_id_image_id_uindex on image_tags image_tags_4  (cost=0.43..426.29 rows=143 width=4) (actual time=0.021..0.130 rows=293 loops=1)
                                Index Cond: (tag_id = 386)
                                Heap Fetches: 72
  ->  Index Scan using images_pkey on images  (cost=0.42..8.30 rows=1 width=636) (actual time=0.006..0.006 rows=1 loops=5)
        Index Cond: (id = tags.image_id)
```
Costs drop yet again from 9.88..8160.16 to 0.85..8114.32.  My image_tags size has increased to 1,996,725, so there is a marginal cost increase as a result of that.

Every subquery before there was this:

```
    ->  Subquery Scan on "*SELECT* 1"  (cost=9.54..537.90 rows=143 width=8) (actual time=51.069..198.055 rows=221 loops=1)
        ->  Bitmap Heap Scan on image_tags  (cost=9.54..536.47 rows=143 width=4) (actual time=51.067..197.969 rows=221 loops=1)
                Recheck Cond: (tag_id = 2082)
                Heap Blocks: exact=232
                ->  Bitmap Index Scan on images_tags_tag_id_idx  (cost=0.00..9.50 rows=143 width=0) (actual time=40.719..40.719 rows=258 loops=1)
                    Index Cond: (tag_id = 2082)
```

The main difference is that every subquery lookup is using the tag_id,image_id index, as expected.

The other difference is that the new explain does not have any Bitmap Heap Scan steps between the Subquery Scan and the Bitmap Index Scan. The bitmap heap scan is where it was performing the image_id matching, it would load the image_ids into memory and do the lookup there. The new index allows the database to use the index immediately for matching both columns. This should result in marginally less memory usage for the overall query, depending on the number of records returned by the tag scans.

The index also gets used by the negative-only JOIN/NOT IN query.

I ran some explains on the other INNER JOIN variant of the search query and its costs also dropped, similarly removing the Bitmap Heap Scans. 

I continue to regard the INCLUDE/EXCLUDE pattern as the best performing, but this index change would benefit any of the implementations.

**Comment by sanmadjack** [comment]
I had another idea this morning, a subquery that performs joins only among the image_tags table, like this:

```
SELECT images.*
FROM images INNER JOIN (
    SELECT it.image_id FROM image_tags it
     INNER JOIN image_tags it2 ON it2.image_id = it.image_id AND it2.tag_id = 6619
     INNER JOIN image_tags it3 ON it3.image_id = it.image_id AND it3.tag_id IN (6613, 6619)
     INNER JOIN image_tags it4 ON it4.image_id = it.image_id AND it4.tag_id IN (14318, 7346, 8631, 26382, 30742, 13562, 5267, 9558, 26352, 5103, 13149, 2083)
     LEFT JOIN image_tags it5 ON it5.image_id = it.image_id AND it5.tag_id IN (386)
    WHERE it.tag_id = 2082 AND it5.image_id IS NULL
    GROUP BY it.image_id
) tags ON tags.image_id = images.id;
```

The execution plan change is amazing:

```
Nested Loop  (cost=2.56..901.64 rows=1 width=636) (actual time=0.564..0.917 rows=5 loops=1)
  ->  Group  (cost=2.14..893.18 rows=1 width=4) (actual time=0.551..0.878 rows=5 loops=1)
        Group Key: it.image_id
        ->  Nested Loop Anti Join  (cost=2.14..893.18 rows=1 width=4) (actual time=0.550..0.873 rows=5 loops=1)
              ->  Nested Loop  (cost=1.71..878.90 rows=1 width=4) (actual time=0.540..0.847 rows=5 loops=1)
                    Join Filter: (it2.image_id = it3.image_id)
                    ->  Nested Loop  (cost=1.28..869.15 rows=1 width=12) (actual time=0.525..0.793 rows=5 loops=1)
                          ->  Merge Join  (cost=0.85..853.30 rows=1 width=8) (actual time=0.510..0.742 rows=5 loops=1)
                                Merge Cond: (it.image_id = it2.image_id)
                                ->  Index Only Scan using image_tags_tag_id_image_id_uindex on image_tags it  (cost=0.43..426.29 rows=143 width=4) (actual time=0.029..0.264 rows=219 loops=1)
                                      Index Cond: (tag_id = 2082)
                                      Heap Fetches: 202
                                ->  Index Only Scan using image_tags_tag_id_image_id_uindex on image_tags it2  (cost=0.43..426.29 rows=143 width=4) (actual time=0.019..0.387 rows=476 loops=1)
                                      Index Cond: (tag_id = 6619)
                                      Heap Fetches: 439
                          ->  Index Only Scan using image_tags_image_id_tag_id_key on image_tags it4  (cost=0.43..15.84 rows=1 width=4) (actual time=0.007..0.009 rows=1 loops=5)
                                Index Cond: (image_id = it2.image_id)
                                Filter: (tag_id = ANY ('{14318,7346,8631,26382,30742,13562,5267,9558,26352,5103,13149,2083}'::integer[]))
                                Rows Removed by Filter: 3
                                Heap Fetches: 21
                    ->  Index Only Scan using image_tags_tag_id_image_id_uindex on image_tags it3  (cost=0.43..9.73 rows=1 width=4) (actual time=0.008..0.009 rows=1 loops=5)
                          Index Cond: ((tag_id = ANY ('{6613,6619}'::integer[])) AND (image_id = it4.image_id))
                          Heap Fetches: 5
              ->  Index Only Scan using image_tags_tag_id_image_id_uindex on image_tags it5  (cost=0.43..7.36 rows=1 width=4) (actual time=0.004..0.004 rows=0 loops=5)
                    Index Cond: ((tag_id = 386) AND (image_id = it.image_id))
                    Heap Fetches: 0
  ->  Index Scan using images_pkey on images  (cost=0.42..8.44 rows=1 width=636) (actual time=0.006..0.006 rows=1 loops=5)
        Index Cond: (id = it.image_id)

```
Everything about this looks better. The costs have plummeted, nothing is above 1000, it's doing more efficient key lookups to filter out more data on each subsequent subquery, and it's still putting off the image table scan until the end. This is now my recommended query.

**Comment by sanmadjack** [comment]
On a related note, I found that the queries for related tags were running very inefficiently. I was artificially inflating the raw amount of tag data by adding several tags to all images, which I would then use as search terms to see if I could somehow cause my current proposed query to have performance issues (it didn't, handled it just fine). I noticed that the image pages were taking up to 30 seconds to load, and I traced it down to the related tags query. This is the query I found in the code:

```
	SELECT t3.tag AS tag, t3.count AS calc_count, it3.tag_id
			FROM
				image_tags AS it1,
				image_tags AS it2,
				image_tags AS it3,
				tags AS t1,
				tags AS t3
			WHERE
				it1.image_id=:image_id
				AND it1.tag_id=it2.tag_id
				AND it2.image_id=it3.image_id
				AND t1.tag != 'tagme'
				AND t3.tag != 'tagme'
				AND t1.id = it1.tag_id
				AND t3.id = it3.tag_id
			GROUP BY it3.tag_id, t3.tag, t3.count
			ORDER BY calc_count DESC
			LIMIT :tag_list_length
```

At a glance I didn't think there should be an issue, it's the old join syntax but it's still joins. So, I ran it through explain to see what was up. 
```

Limit  (cost=2914668.25..2914669.85 rows=15 width=19) (actual time=18389.593..19044.905 rows=15 loops=1)
  ->  Group  (cost=2914668.25..6986064.99 rows=38191779 width=19) (actual time=18389.592..19033.791 rows=15 loops=1)
        Group Key: t3.count, it3.tag_id, t3.tag
        ->  Gather Merge  (cost=2914668.25..6747366.38 rows=31826482 width=19) (actual time=18389.588..19044.840 rows=43 loops=1)
              Workers Planned: 2
              Workers Launched: 2
              ->  Group  (cost=2913668.22..3072800.63 rows=15913241 width=19) (actual time=15585.209..16715.395 rows=391 loops=3)
                    Group Key: t3.count, it3.tag_id, t3.tag
                    ->  Sort  (cost=2913668.22..2953451.33 rows=15913241 width=19) (actual time=15585.205..16301.687 rows=2453454 loops=3)
                          Sort Key: t3.count DESC, it3.tag_id, t3.tag
                          Sort Method: external merge  Disk: 228232kB
                          Worker 0:  Sort Method: external merge  Disk: 249680kB
                          Worker 1:  Sort Method: external merge  Disk: 331920kB
                          ->  Hash Join  (cost=70355.97..357454.27 rows=15913241 width=19) (actual time=3848.744..8147.206 rows=8459405 loops=3)
                                Hash Cond: (it3.image_id = it2.image_id)
                                ->  Hash Join  (cost=1458.64..58502.71 rows=2379540 width=23) (actual time=81.479..955.422 rows=1684972 loops=3)
                                      Hash Cond: (it3.tag_id = t3.id)
                                      ->  Parallel Seq Scan on image_tags it3  (cost=0.00..50796.98 rows=2379598 width=8) (actual time=0.573..311.385 rows=1906726 loops=3)
                                      ->  Hash  (cost=947.83..947.83 rows=40865 width=19) (actual time=62.972..62.973 rows=40866 loops=3)
                                            Buckets: 65536  Batches: 1  Memory Usage: 2633kB
                                            ->  Seq Scan on tags t3  (cost=0.00..947.83 rows=40865 width=19) (actual time=0.689..14.490 rows=40866 loops=3)
                                                  Filter: ((tag)::text <> 'tagme'::text)
                                                  Rows Removed by Filter: 1
                                ->  Hash  (cost=582.72..582.72 rows=4163889 width=4) (actual time=3746.040..3746.040 rows=3331881 loops=3)
                                      Buckets: 131072  Batches: 64  Memory Usage: 2872kB
                                      ->  Nested Loop  (cost=1.15..582.72 rows=4163889 width=4) (actual time=15.413..2034.549 rows=3331881 loops=3)
                                            Join Filter: (it1.tag_id = it2.tag_id)
                                            ->  Nested Loop  (cost=0.72..115.36 rows=9 width=8) (actual time=8.491..17.692 rows=9 loops=3)
                                                  ->  Index Only Scan using image_tags_image_id_tag_id_key on image_tags it1  (cost=0.43..40.57 rows=9 width=4) (actual time=7.700..8.695 rows=10 loops=3)
                                                        Index Cond: (image_id = 697099)
                                                        Heap Fetches: 30
                                                  ->  Index Scan using tags_pkey on tags t1  (cost=0.29..8.31 rows=1 width=4) (actual time=0.895..0.895 rows=1 loops=30)
                                                        Index Cond: (id = it1.tag_id)
                                                        Filter: ((tag)::text <> 'tagme'::text)
                                                        Rows Removed by Filter: 0
                                            ->  Index Only Scan using image_tags_tag_id_image_id_uindex on image_tags it2  (cost=0.43..32.35 rows=1566 width=8) (actual time=1.126..165.176 rows=370209 loops=27)
                                                  Index Cond: (tag_id = t1.id)
                                                  Heap Fetches: 9993009

```

The key takeaway here:

                          Worker 0:  Sort Method: external merge  Disk: 249680kB
                          Worker 1:  Sort Method: external merge  Disk: 331920kB

It was manually joining the tables, ignoring indexing for a good chunk of the work. According to that plan it was using hundreds of MB of disk space to run the query. That's bigger than the actual table is. The total cost for the query (which granted, my data set right now is worst-case scenario) is almost 3 million.

So, I switched it over to modern join syntax:

```
SELECT t3.tag AS tag, t3.count AS calc_count, it3.tag_id
            FROM image_tags AS it1 -- Starting image's tags                
                INNER JOIN tags AS t1 ON t1.id = it1.tag_id AND t1.tag NOT LIKE 'tagme%'
                -- Get images with the same tags as the starting image
                INNER JOIN image_tags AS it2 ON it1.tag_id=it2.tag_id
                -- Get the tags from those other images except the same as the starting tags 
                INNER JOIN image_tags AS it3 ON it2.image_id=it3.image_id 
                    AND it3.tag_id not in (SELECT tag_id FROM image_tags WHERE image_id = it1.image_id)
                INNER JOIN tags AS t3 ON t3.id = it3.tag_id AND t3.tag NOT LIKE 'tagme%' 
			WHERE
				it1.image_id=:image_id
			GROUP BY it3.tag_id, t3.tag, t3.count
			ORDER BY calc_count DESC
			LIMIT :tag_list_length
```

And now it's returning in about 700 ms. Still longer than I might like, but considering how much data it's dealing with it's handling it like a champ. It's also now making full use of the tag_id, image_id index. I've updated both of the related tags blocks in this manner. I'm satisfied enough with the performance improvements for this and the tag queries that I'm going to go ahead and get a pull request together.

**Comment by sanmadjack** [comment]
This appears to be a duplicate of #618
